packages/core/http/core-http-router-server-internal/src/request.ts

```diff
@@ -73,13 +82,49 @@ export class CoreKibanaRequest<
     withoutSecretHeaders: boolean = true
   ) {
     let requestParts: { params: P; query: Q; body: B };
-    if (routeSchemas === undefined || isFakeRawRequest(req)) {
-      requestParts = { query: {} as Q, params: {} as P, body: {} as B };
+
+    if (isFakeRawRequest(req)) {
```

Comment:
Hm, this is a tricky one. We actually construct the request object many times over. Once for every registered lifecycle handler and the "real" handler which is the only time we call it with validation schemas and hence a value for `ELASTIC_HTTP_VERSION_QUERY_PARAM`.

My concern with this approach is that we will be passing down `ELASTIC_HTTP_VERSION_QUERY_PARAM` to handlers whether they are versioned or not... Only the versioned router knows about this value and will clean it up. I don't think this is a massive issue but ideally we would avoid it.

Just for my understanding, what is the current issue with only getting the versioned route's security in (`getRouteInfo`) just before the real handler is called?

Reply:
Not sure that I follow, we get security in `getRouteInfo` for versioned route already [here](https://github.com/elastic/kibana/pull/191973/files#diff-c1f4a8c013c6ff9c5170803322414cb002f70331b23e45c55f075258c678410aR308). For the versioned route the security is a function, so we can obtain the correct security for specific version that we get on request, but the query param gets stripped out before in`CoreKibanaRequest.from`. Am I missing smth?

Reply:
I see that [requestHandler](https://github.com/elastic/kibana/blob/main/packages/core/http/core-http-router-server-internal/src/versioned_router/core_versioned_route.ts#L125) gets raw request with those params available, but the authz check happens in on post auth hook before that handler is called and we need that security config be available there for the requested version

Reply:
> I see that [requestHandler](https://github.com/elastic/kibana/blob/main/packages/core/http/core-http-router-server-internal/src/versioned_router/core_versioned_route.ts?rgh-link-date=2024-09-20T09%3A41%3A55Z#L125) gets raw request with those params available, but the authz check happens in on post auth hook before that handler is called and we need that security config be available there for the requested version

Yea, getting config propagated to where you need it takes a bit of fiddling.

Reply:
>...the authz check happens in on post auth hook before that handler 

I see. Yeah the "special" query param is not stripped, it's just not passed down because we couldn't run validation. It's still on the "raw request".

What do you think about something like:

```
   return isRouteSecurityGetter(securityConfig) ? securityConfig(this[requestSymbol]) : securityConfig
```

That way, the full, unvalidated  raw request should be passed to the security getter. Then we just need to make sure the versioned router's `getSecurity` function can read an object that matches both `KibanaRequest` and Hapi's `Request`. We can update the read version utils to something like:

<details>

<summary>diff</summary>

```
diff --git a/packages/core/http/core-http-router-server-internal/src/versioned_router/route_version_utils.ts b/packages/core/http/core-http-router-server-internal/src/versioned_router/route_version_utils.ts
index f8645f947be..1e1d67203b5 100644
--- a/packages/core/http/core-http-router-server-internal/src/versioned_router/route_version_utils.ts
+++ b/packages/core/http/core-http-router-server-internal/src/versioned_router/route_version_utils.ts
@@ -63,13 +63,18 @@ export function removeQueryVersion(request: Mutable<KibanaRequestWithQueryVersio
   delete request.query[ELASTIC_HTTP_VERSION_QUERY_PARAM];
 }
 
-function readQueryVersion(request: KibanaRequest): undefined | ApiVersion {
+interface RequestLike {
+  headers: KibanaRequest['headers'];
+  query: KibanaRequest['query'];
+}
+
+function readQueryVersion(request: RequestLike): undefined | ApiVersion {
   const version = get(request.query, ELASTIC_HTTP_VERSION_QUERY_PARAM);
   if (typeof version === 'string') return version;
 }
 /** Reading from header takes precedence over query param */
 export function readVersion(
-  request: KibanaRequest,
+  request: RequestLike,
   isQueryVersionEnabled?: boolean
 ): undefined | ApiVersion {
   const versions = request.headers?.[ELASTIC_HTTP_VERSION_HEADER];
```

</details>


One optimisation (up to you) to avoid running `readVersion` code twice we can "promote" the read version logic to the `Request` object and include the version in the route info here: https://github.com/elastic/kibana/blob/c17945191ddef3a8cb80ab900049033857d3b73f/packages/core/http/core-http-router-server-internal/src/request.ts#L244

That way the `getSecurity` getter can take a version... But I don't think it's a huge deal.

Reply:
Added in f55463624ea2681e13381dd397f00493cc2a9d49